nextcloud: init at 10.0.0

schneefux opened the pull request:
###### Motivation for this change

Based on the OwnCloud package, added the latest NextCloud 9.0 and NextCloud 10.0 packages.
Added an up-to-date patch that implements the possibility to specify a configuration file via environment variables, as seen [here](https://github.com/NixOS/nixpkgs/blob/staging/pkgs/servers/owncloud/default.nix#L18).

To use with nginx:

```
…
root ${pkgs.nextcloud};
…
location ~ \.php(?:$|/) {
  …
  fastcgi_param NEXTCLOUD_CONFIG_DIR '/etc/nextcloud';
  …
}
…
location ^~ /apps {
  alias /var/www/nextcloud-apps;
}
…
```

Additional NextCloud configuration options needed:

``` php
…
'datadirectory' => '/var/www/nextcloud-data',
'apps_paths' =>
 array(
  0 =>
   array(
    'path' => '/var/www/nextcloud-apps',
    'url' => '/apps',
    'writable' => true,
  ),
),
…
```
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [x] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

/cc @matejc

ericsagnes reviewed (commented):
This is coming from the owncloud derivation, but the `versiona = version` could be replaced by a more idiomatic `inherit version;` and changing all `versiona` to `version`.

rasendubi commented on pkgs/servers/nextcloud/default.nix:
```diff
@@ -0,0 +1,48 @@
+
+in {
+
+  nextcloud90 = common {
```
Why we need multiple versions?

rasendubi commented on pkgs/servers/nextcloud/default.nix:
```diff
@@ -0,0 +1,48 @@
+{ stdenv, fetchurl, fetchpatch }:
+let
+  common = { version, sha256 } @ args: stdenv.mkDerivation (rec {
```
why `@ args`?

rasendubi replied:
you don't need parenthesis around `rec { }`

rasendubi commented on pkgs/servers/nextcloud/default.nix:
```diff
@@ -0,0 +1,48 @@
+        find . -maxdepth 1 -execdir cp -r '{}' $out \;
+      '';
+
+    meta = {
```
Would like to be a package maintainer?

schneefux committed:
nextcloud: init at 10.0.0

schneefux committed:
maintainers: add schneefux

schneefux commented:
I've rewritten the package based on your suggestions.

> Why we need multiple versions?

`owncloud.nix` did it. I removed it, now it's much simpler.

> Would like to be a package maintainer?

Well, why not :)

rasendubi reviewed (approved):
:+1:
